Fix/flex based layout for side menu and main content

wielinde opened the pull request:
https://community.openproject.com/projects/openproject/work_packages/25677/activity

I am pretty sure that this PR will trigger other bugs to come up due to its complexity.

For reviewing/testing this PR I found that at least the following combinations have to be tested:

All test cases need to be checked for:

- Side menu expanded
- Side menu collapsed
- No menu (i.e. global stuff)
- Zen Mode (in WPs)

Screen sizes:

- Large desktop (that allows for displaying split screen and timelines)
- Medium desktop (so that in WP list the activity and relations are collapsed)
- Mobile (Side menu should not have a full column and instead overlap #main)

Specials stuff:
- The WP List / Timelines work in a different lay outing mode (position: absolute)
- From that exception medium desktops have an exception from that exception (setting position: relative again) when showing a WP in full screen (so not in split view).

Breadcrumbs:

There are multiple types of breadcrumbs that had all some complex lay outing rules that now should get simplified.

wielinde committed:
flex based layout for side menu and main content

wielinde committed:
add an exception for the height of work package table containers for the summary view, as there are multiple tables in one view.

wielinde commented:
@RobinWagner Could you please have a look at this PR whenever you have time and a PP is ready? Testing this PR is a hard one.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ddefa642d253b250cef25fff4676271f9e637645) (deployed commit: 0d5123bf4e14cd320ebbedac52afb5526ea6aac9).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/19a045d55ddadcb9ef1383e0c37b3f1ea34d9bad) (deployed commit: 0d5123bf4e14cd320ebbedac52afb5526ea6aac9).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/83b2e0de82cd070fce546affa6c72c96036e68d6) (deployed commit: 0d5123bf4e14cd320ebbedac52afb5526ea6aac9).

wielinde committed:
Fixing mobile WP list height. It was 0.

wielinde committed:
I do not see a good reason why having a different header size for mobile has any advantage. I start removing this for the sake of reducing complexity

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ca5228e149d1fa0f29e3017744abf8cec8ed39eb) (deployed commit: 9a0a2765b921f2c7bfa3abef02448d177c409aa7).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/f112752e92502968b9df7363161bb665c85d71a5) (deployed commit: 62aaebd3b27a5459eeb28cb043eb84b6494a5c66).

wielinde committed:
Fix tiny gap below main when right content-wrapper was taller than the menu

wielinde committed:
Unrelated but I believe it is right to use SASS variables if we have them already defined.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/f380f2cb7b09a14c253d00bdbe530b9fe8768d2a) (deployed commit: b6309cf361ba0951e9bb30ed1a88b6b72dbb1f5f).

RobinWagner commented:
Hi @wielinde,

here are the issues I found:

- The subject column gets truncated too much:
![bug_subject-column](https://user-images.githubusercontent.com/4333337/27685570-15e3c0b8-5ccf-11e7-820b-6edab7e45cab.PNG)

- Sidebar icon jumps when collapsing (minor issue):
![bug_icon-side-menu](https://user-images.githubusercontent.com/4333337/27685581-20bdcc68-5ccf-11e7-99e5-5931081c5c5c.gif)

- On responsive / smaller screensizes the side container seems to be too small:
![bug_responsive-screen](https://user-images.githubusercontent.com/4333337/27685587-2af0ddba-5ccf-11e7-8fdd-2a4d6689e5ba.PNG)
![bug_responsive-screen](https://user-images.githubusercontent.com/4333337/27685616-581cdea6-5ccf-11e7-98ab-90bcc0ba8c8e.gif)

- Relations in timelines get removed when updating (not sure if this depends on this PR):
![bug_relation-lost](https://user-images.githubusercontent.com/4333337/27685604-46dc4988-5ccf-11e7-9c18-f336165adb1d.gif)

Otherwise it looks good to me.

wielinde commented:
@RobinWagner 

Thank you for the reports. Let me answer bug by bug.

_The subject column gets truncated too much_
I cannot reproduce that. Could you help me to reproduce it? However, I bet that this bug existed before as I cannot think of anything where I touched table internal rules. I cannot really say how often that occurs and thus cannot say how important a fix would be.

_Sidebar icon jumps when collapsing (minor issue)_
Thank you. That is a bug that existed before. It might get fixed/obsolete when we work on this feature: https://community.openproject.com/projects/openproject/work_packages/25496/activity

_On responsive / smaller screensizes the side container seems to be too small_
Yes, I see what you mean. This bug would get visible on big mobiles (not on iphones I believe). However, that bug existed before this PR and should get fixed when we optimize the menu for mobile.... so IMHO same feature: https://community.openproject.com/projects/openproject/work_packages/25496/activity

_Relations in timelines get removed when updating (not sure if this depends on this PR)_
I am not sure if I observed the same bug. What I see is that after adding a relation (before reloading) other relations disappear. After reloading all are visible again. However, that is not related to this PR and should go into a separate bug report. I believe that bug is somewhat important as it has a high potential for user frustration.

Thank you again and please let me know, if I missed something.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/f83483d8b31d7903a184f5fdc74e29a6587497c8) (deployed commit: b6309cf361ba0951e9bb30ed1a88b6b72dbb1f5f).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/97fb131d62926a6d61bb2d2b3d8f8a0d4468c813) (deployed commit: b6309cf361ba0951e9bb30ed1a88b6b72dbb1f5f).

oliverguenther commented:
LGTM. I'd like to give @HDinger a chance to look at this also. Free to merge otherwise, the failing test is unrelated (happens on dev as well).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/848fb7851e551da9503434feb07136e99955c40c) (deployed commit: b6309cf361ba0951e9bb30ed1a88b6b72dbb1f5f).

HDinger commented:
Looks good. However I found some issues:

- On mobile pages there is a horizontal scrollbar within the side menu.
![bildschirmfoto 2017-07-07 um 09 03 26](https://user-images.githubusercontent.com/7457313/27947022-84f4e7ea-62f4-11e7-85e2-f3ae09e6d5a5.png)

- In Safari the WP table is too much intended.
![bildschirmfoto 2017-07-07 um 09 02 51](https://user-images.githubusercontent.com/7457313/27947104-cac90530-62f4-11e7-85e1-452ba2475c9f.png)
 
- On pages with no side bar but a breadcrumb the indentation is wrong.
![bildschirmfoto 2017-07-07 um 09 34 06](https://user-images.githubusercontent.com/7457313/27947768-8392c400-62f7-11e7-8226-bb8e2b094ed8.png)

- I guess this is the same issue Robin already mentioned (?): When the side menu does not fit into the page, the scrolling behavior is applied to the whole page, not to the side menu itself.
![bildschirmfoto 2017-07-07 um 09 03 34](https://user-images.githubusercontent.com/7457313/27947221-4250866e-62f5-11e7-8cf7-7e260ea6ab26.png)

Further there are two issues which we probably won't change now, since they are subjective opinions. I guess Niels has to decide whether this bothers him or not: 

- The mobile header hight was increased. However I can remember that we made it smaller on purpose to gain more space for the mobile view.
- A similar issue occurs for the breadcrumb. It has now a lot of space. I can imagine that it is too much.
![bildschirmfoto 2017-07-07 um 09 07 56](https://user-images.githubusercontent.com/7457313/27947455-228cba18-62f6-11e7-9eff-f53dc4461719.png)

HDinger reviewed (changes requested):
Before we merge the Safari bug, the breadcrumb indentation and the horizontal scroll bar bug need to be fixed.

HDinger commented:
As discussed I will take care of the bugs I found.

HDinger committed:
Fix bugs with new flex layout

HDinger commented:
@oliverguenther @wielinde I fixed the mandatory bugs, so for me this is free to go. Maybe you can have a short look at the changes I made.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/64c2884c78ce64f207ad5f3746b86768fe263b3e) (deployed commit: c07a261d96a5f0645949f15e2a9938048902f6f4).

wielinde committed:
getting rid of horizontal scrollbar in main menu on mobiles.

wielinde committed:
Undo this this change introduced by @HDinger two commits ago as it introduces a bug: In full screen view of a work package the main menu does not reach the bottom which is actually the main purpose of this PR.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/24449ecec66f43092009b13e45f8c6333924d967) (deployed commit: cf500c520b56d04f7b9d52d7fec2b336cbb0b58d).

wielinde commented:
@RobinWagner Could you please test this again?

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/d20999ca5783acca27e25924bd159d428b6af0cf) (deployed commit: cf500c520b56d04f7b9d52d7fec2b336cbb0b58d).

RobinWagner commented:
Hi @wielinde,

I could not test the Safari bugs that Henriette mentioned (since I don't have a Mac).

It generally looks good.

I noticed a couple of other smaller bugs:

- The scroll bar is missing on several admin pages (e.g. projects, users). This makes it difficult for users with smaller screens to use the pages
![bug_scrolling1](https://user-images.githubusercontent.com/4333337/28074583-3b106100-6659-11e7-8165-a8adfaecbab1.PNG)
![bug_scrolling2](https://user-images.githubusercontent.com/4333337/28074591-3ee7181e-6659-11e7-85f4-a5e8933e1659.PNG)

- I get an internal error on the calendar page (probably unrelated to this PR):
![bug_calendar-internal-error](https://user-images.githubusercontent.com/4333337/28074713-8dedf32e-6659-11e7-8c8b-fff33b336588.gif)

Otherwise, it looks good 👍

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/d5b12d76d91c15a567d53b4a11a790a32a492dda) (deployed commit: d7a5552671a6510d8d7c00cfd78e6c2baf2c7245).

wielinde committed:
Fixing side menu width

[ci skip]

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ce3f11611d083eb58c3f163cd4ce4ca9f582c9b4) (deployed commit: ffe14e06bc26b4ff11f2e8b1c9e0742fffdade41).

wielinde committed:
with a collapsed main menu the right border of its toggle was missing

[ci skip]

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ee347cbf5a7c4df02be8b4b9482797201c6786c1) (deployed commit: 18f1e834732040c2200e4875800072185fb5cad8).

wielinde committed:
Fixing overflow settings for the now flex based #main children

[ci skip]

wielinde commented:
@RobinWagner Wow, the scrollbar bug was a really interesting one. Thank you. I think I fixed it.

The calendar bug is not related. I believe it to be something else.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ece28d7280b95b43ecddadab0cb7a2f3384dead0) (deployed commit: f6f7c7329935a79b2207143a22110d5a64dd1009).

RobinWagner commented:
@wielinde This works properly now. Thanks! :)

I noticed a couple of bugs but they already exist on the current production system, so they should be unrelated:
https://community.openproject.com/projects/openproject/work_packages/25785/activity
https://community.openproject.com/projects/openproject/work_packages/25786/activity

I don't know if the Safari bugs have been fixed since I cannot test this on my system.
But otherwise, it looks good to me!

Great job 👍

wielinde committed:
Fix main-menu scrolling/overflow for Safari

[ci skip]

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/cf1dd00104563695ebcf5116787a1ddcb54632c7) (deployed commit: 72e60c3a8fdb98120284303f3619423dd660b16c).

wielinde committed:
Fixing mobile main manu

Fix for
https://community.openproject.com/projects/openproject/work_packages/257
85/activity

„On mobile, side navigation is cut off if content is shorter than side
navigation“

[ci skip]

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/3fa9f7bbef0dbbdd3f6666b9c3f3f74a4bc4db60) (deployed commit: f94874c5b02303429aee33c7d14f3ebeec8210e3).

wielinde committed:
Remove obsolete background colouring hacks as we now have proper flex based columns to color

[ci skip]

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/b78c1faff2f5a0ff0ff717c90854ed84be9993a4) (deployed commit: aeae0894fc37d8436a9409b0cabfcaad5433ca59).
